### PR TITLE
Improve Telegram final exit msg

### PR DESCRIPTION
**Pull request by xmatthias**
## Summary

The final exit message is - in some cases - not displaying the correct final profit (it'll display the "last leg" profit instead).
This PR will address this.

@stash86 as you've been the one to report this - would you mind giving this a spin? 
The way i've tested it is by removing the last order from your trade and reopening the trade - and then selling the rest again - which then created the below image.

## Quick changelog

- Improve final exit message in case of multiple exits

## What's new?

image

### CODE REVIEW

**Comment by perkmeister** [comment]
> The final exit message is - in some cases - not displaying the correct final profit (it'll display the "last leg" profit instead).
> This PR will address this.

Isn't that what I thought might be happening 6 months ago?

**Comment by stash86** [comment]
It works as intended (the last exit message), but I find a new bug(?). Sometimes my partial exit labeled correctly as partial exit, hence I got the sub and cumulative profits, but sometimes it labeled as normal exit despite it being partial exit, but then I don't get the sub and final profits.

image

Sometimes these is what I got
image

Final exit
image

**Comment by xmatthias** [comment]
> Isn't that what I thought might be happening 6 months ago?

I don't really remember that - maybe, maybe not - however apparently, that wasn't shown / explained in a way to make yourself understood. this PR is not an "my opinion on this changed" - but "the evidence of what's happening showed wrong behavior"...
 (my strategies usually do not use partial exits, hence it's not something i run across myself frequently).

---

@stash86 i'm not sure i understand the middle screenshot ...  (as in - it's lacking details).

it seems like the first partial exit ? 
in which case, profit and cumulative profit would be identical ... so it's not shown separtely

**Comment by stash86** [comment]
Yes, for the middle screenshot, the top exit was the first partial exit. But notice the second partial exit, it shown as full exit despite being partial exit

**Comment by xmatthias** [comment]
> Yes, for the middle screenshot, the top exit was the first partial exit. But notice the second partial exit, it shown as full exit despite being partial exit

What makes you believe that (i don't see what you mean ...) ? 

there's no "Final Profit" line, which i'd expect from a "full exit" in case of a multi-exit strategy.

**Comment by stash86** [comment]
Because it says "Exited BTC/USDT" instead of "Partially exited BTC/USDT". It's just labelling, but why the different labels?

**Comment by stash86** [comment]
Okay, this is the full breakdown of the bug. Expect a long texts

This is initial entry

```
✓ Binance (dry): Longed BTC/USDT (#1)
Amount: 0.00034000
Open Rate: 42869.56000000
Total: (14.576 USDT, 14.59 USD)
```

This is 2nd entry
```
✓ Binance (dry): Longed BTC/USDT (#1)
Amount: 0.00033000
Open Rate: 42870.27000000
Total: (28.723 USDT, 28.752 USD)
```

This is 1st partial exit (notice the remaining part at the bottom)
```
❌ Binance (dry): Exited BTC/USDT (#1)
Cumulative Profit: -0.19% (loss: -0.02751932 USDT / -0.028 USD)
Enter Tag: None
Exit Reason: partial_exit
Direction: Long
Amount: 0.00033000
Open Rate: 42869.90970149
Exit Rate: 42872.26000000
Min Profit: 0.00%
Max Profit: 0.01%

Remaining: (14.576 USDT, 14.59 USD)
```

This is 3rd entry
```
✓ Binance (dry): Longed BTC/USDT (#1)
Amount: 0.00034000
Open Rate: 42851.23000000
Total: (29.145 USDT, 29.174 USD)
```

This is 2nd partial exit (notice there is no remaining line at the bottom, I got Duration instead, that indicate `is_sub_trade` at this point is False where it should be True)
```
❌ Binance (dry): Exited BTC/USDT (#1)
Profit: -0.21% (loss: -0.02999089 USDT / -0.030 USD)
Enter Tag: None
Exit Reason: partial_exit
Direction: Long
Amount: 0.00034000
Open Rate: 42860.56985075
Exit Rate: 42858.08000000
Min Profit: -0.04%
Max Profit: 0.01%

Duration: 0:01:04 (1.1 min)
```

This is 4th entry
```
✓ Binance (dry): Longed BTC/USDT (#1)
Amount: 0.00033000
Open Rate: 42868.93000000
Total: (28.719 USDT, 28.748 USD)
```

This is 3rd partial exit (The message format is correct)
```
❌ Binance (dry): Partially exited BTC/USDT (#1)
Sub Profit: -0.16% (loss: -0.02258659 USDT / -0.023 USD)
Cumulative Profit: (-0.08009680 USDT / -0.080 USD)
Enter Tag: None
Exit Reason: partial_exit
Direction: Long
Amount: 0.00033000
Open Rate: 42864.68753620
Exit Rate: 42881.99000000
Min Profit: -0.04%
Max Profit: 0.03%

Remaining: (14.574 USDT, 14.589 USD)
```

To make this short, 4th and 5th part exit messages are also correct.
This is 6th partial exit message, which once again indicate `is_sub_trade` is seen as False which is wrong
```
❌ Binance (dry): Exited BTC/USDT (#1)
Profit: -0.20% (loss: -0.02937417 USDT / -0.029 USD)
Enter Tag: None
Exit Reason: partial_exit
Direction: Long
Amount: 0.00034000
Open Rate: 42881.72179682
Exit Rate: 42881.09000000
Min Profit: -0.04%
Max Profit: 0.06%

Duration: 0:02:43 (2.7 min)
```

Messages for partial exit 7 to 11 are correct.
This is the full exit message
```
❌ Binance (dry): Exited BTC/USDT (#1)
Sub Profit: -0.17% (loss: -0.04915659 USDT / -0.049 USD)
Final Profit: -0.17% (-0.32284202 USDT / -0.323 USD)
Enter Tag: None
Exit Reason: roi
Direction: Long
Amount: 0.00069000
Open Rate: 42926.28423781
Exit Rate: 42940.91000000
Min Profit: -0.04%
Max Profit: 0.17%

Duration: 0:05:03 (5.0 min)
```

So sometimes `is_sub_trade` can be seen as False

**Comment by stash86** [comment]
You can test that using this strat

```
import freqtrade.vendor.qtpylib.indicators as qtpylib
import numpy as np
import talib.abstract as ta
from freqtrade.strategy import IStrategy, informative
from pandas import DataFrame, Series
from typing import Dict, List, Optional, Tuple
from freqtrade.persistence import Trade
from datetime import datetime, timedelta, timezone
import logging

logger = logging.getLogger(__name__)

class test_grind (IStrategy):

    INTERFACE_VERSION = 3

    # ROI table:
    minimal_roi = {
        "0": 1000,
        "5": -1
    }

    # Stoploss:
    stoploss = -0.99

    # Trailing stop:
    trailing_stop = False
    trailing_stop_positive = 0.001
    trailing_stop_positive_offset = 0.01
    trailing_only_offset_is_reached = False

    # Sell signal
    use_exit_signal = False
    exit_profit_only = False
    exit_profit_offset = 0.01
    ignore_roi_if_entry_signal = False

    timeframe = '5m'

    process_only_new_candles = True
    startup_candle_count = 1

    position_adjustment_enable = True
    max_entry_position_adjustment = 100
    max_dca_multiplier = 1

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        
        return dataframe
    
    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        
        dataframe.loc[:, 'enter_long']= 1

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        dataframe.loc[:, 'exit_long'] = 0

        return dataframe

    # This is called when placing the initial order (opening trade)
    def custom_stake_amount(self, pair: str, current_time: datetime, current_rate: float,
                            proposed_stake: float, min_stake: Optional[float], max_stake: float,
                            leverage: float, entry_tag: Optional[str], side: str,
                            **kwargs) -> float:

        # We need to leave most of the funds for possible further DCA orders
        # This also applies to fixed stakes
        return proposed_stake / self.max_dca_multiplier

    def adjust_trade_position(self, trade: Trade, current_time: datetime,
                              current_rate: float, current_profit: float,
                              min_stake: Optional[float], max_stake: float,
                              current_entry_rate: float, current_exit_rate: float,
                              current_entry_profit: float, current_exit_profit: float,
                              **kwargs) -> Optional[float]:

        filled_entries = trade.select_filled_orders(trade.entry_side)
        count_of_entries = trade.nr_of_successful_entries
        
        filled_exits = trade.select_filled_orders(trade.exit_side)
        count_of_exits = trade.nr_of_successful_exits

        if current_time - timedelta(seconds=15) > filled_entries[(count_of_entries - 1)].order_filled_date.replace(tzinfo=timezone.utc):
            if (count_of_exits + 1) == count_of_entries:
                return trade.stake_amount
            elif (count_of_exits + 1) < count_of_entries:
                exit_amount = trade.stake_amount / 2
                return -(exit_amount)

        return None

```

**Comment by stash86** [comment]
Tried on another vps on fresh install, summary is

Partial exit 1-6 : sub_trade is False
Partial exit 7-11 : sub_trade is True

This is log from one of the partial exit
```
2024-01-05 10:28:28,256 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': exit, 'trade_id': 1, 'exchange': 'Binance', 'pair': 'BTC/USDT', 'leverage': 1.0, 'direction': 'Long', 'gain': 'loss', 'limit': 43900.7, 'order_rate': 43900.7, 'order_type': 'market', 'amount': 0.00045, 'open_rate': 43921.39203125, 'close_rate': 43900.7, 'current_rate': 43900.7, 'profit_amount': -0.28252255, 'profit_ratio': -0.00246818, 'buy_tag': None, 'enter_tag': None, 'sell_reason': 'partial_exit', 'exit_reason': 'partial_exit', 'open_date': datetime.datetime(2024, 1, 5, 10, 25, 44, 43052, tzinfo=datetime.timezone.utc), 'close_date': datetime.datetime(2024, 1, 5, 10, 28, 28, 256464, tzinfo=datetime.timezone.utc), 'stake_amount': 39.529252828125, 'stake_currency': 'USDT', 'base_currency': 'BTC', 'fiat_currency': 'USD', 'sub_trade': True, 'cumulative_profit': -0.23369119, 'final_profit_ratio': None, 'is_final_exit': False}
2024-01-05 10:28:28,259 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=1, pair=BTC/USDT, amount=0.00090000, is_short=False, leverage=1.0, open_rate=43921.39203125, open_since=2024-01-05 10:25:44)
2024-01-05 10:28:28,264 - freqtrade.persistence.trade_model - INFO - Updating trade (id=1) ...
2024-01-05 10:28:28,265 - freqtrade.persistence.trade_model - INFO - MARKET_SELL has been fulfilled for Trade(id=1, pair=BTC/USDT, amount=0.00090000, is_short=False, leverage=1.0, open_rate=43921.39203125, open_since=2024-01-05 10:25:44).
2024-01-05 10:28:28,269 - freqtrade.wallets - INFO - Wallets synced.
2024-01-05 10:28:28,271 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': exit_fill, 'trade_id': 1, 'exchange': 'Binance', 'pair': 'BTC/USDT', 'leverage': 1.0, 'direction': 'Long', 'gain': 'loss', 'limit': 43900.7, 'order_rate': 43900.7, 'order_type': '', 'amount': 0.00045, 'open_rate': 43921.39203125, 'close_rate': 43900.7, 'current_rate': None, 'profit_amount': -0.04883136, 'profit_ratio': -0.00246818, 'buy_tag': None, 'enter_tag': None, 'sell_reason': 'partial_exit', 'exit_reason': 'partial_exit', 'open_date': datetime.datetime(2024, 1, 5, 10, 25, 44, 43052, tzinfo=datetime.timezone.utc), 'close_date': datetime.datetime(2024, 1, 5, 10, 28, 28, 271555, tzinfo=datetime.timezone.utc), 'stake_amount': 19.7646264140625, 'stake_currency': 'USDT', 'base_currency': 'BTC', 'fiat_currency': 'USD', 'sub_trade': False, 'cumulative_profit': -0.28252255, 'final_profit_ratio': None, 'is_final_exit': False}
```
Notice that `sub_trade` is True on `exit` message, but became False on `exit_fill` message

**Comment by xmatthias** [comment]
Nice finds - please give the updated version a try ... ;) it seems to behave well now, at least with the strategy provided.

**Comment by stash86** [comment]
Now I find another bug

Take this for example
This is the first partial exit `exit` message
```
❌ Binance (dry): Partially exiting BTC/USDT (#3)
Unrealized Sub Profit: -0.20% (loss: -0.03887343 USDT / -0.039 USD)
Cumulative Profit: (0.00000000 USDT / 0.000 USD)
```

This is the `exit fill` message
```
❌ Binance (dry): Partially exited BTC/USDT (#3)
Cumulative Profit: -0.20% (loss: -0.03887343 USDT / -0.039 USD)
Cumulative Profit: (-0.03887343 USDT / -0.039 USD)
```

This is where the "bug" started. This is the 2nd partial exit `exit` message
```
❌ Binance (dry): Partially exiting BTC/USDT (#3)
Unrealized Sub Profit: -0.13% (loss: -0.06352187 USDT / -0.064 USD)
Cumulative Profit: (-0.03887343 USDT / -0.039 USD)
```

This is the `exit fill` message
```
❌ Binance (dry): Partially exited BTC/USDT (#3)
Sub Profit: -0.13% (loss: -0.02464844 USDT / -0.025 USD)
Cumulative Profit: (-0.06352187 USDT / -0.064 USD)
```

As you can see, the absolute number on `Unrealized Sub Profit` in the `exit` message is actually absolute number for `Unrealized cumulative profit`

This is 3rd partial exit `exit`
```
❌ Binance (dry): Partially exiting BTC/USDT (#3)
Unrealized Sub Profit: -0.17% (loss: -0.09569611 USDT / -0.096 USD)
Cumulative Profit: (-0.06352187 USDT / -0.064 USD)
```

This is the `exit fill`
```
❌ Binance (dry): Partially exited BTC/USDT (#3)
Sub Profit: -0.17% (loss: -0.03217424 USDT / -0.032 USD)
Cumulative Profit: (-0.09569611 USDT / -0.096 USD)
```

**Comment by stash86** [comment]
This is the logs for the 2nd partial exit
```
2024-01-05 23:34:56,674 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': exit, 'trade_id': 3, 'exchange': 'Binance', 'pair': 'BTC/USDT', 'leverage': 1.0, 'direction': 'Long', 'gain': 'loss', 'limit': 44140.95, 'order_rate': 44140.95, 'order_type': 'market', 'amount': 0.00044, 'open_rate': 44108.719508900394, 'close_rate': 44140.95, 'current_rate': 44140.95, 'profit_amount': -0.06352187000000001, 'profit_ratio': -0.00126876, 'buy_tag': None, 'enter_tag': None, 'sell_reason': 'partial_exit', 'exit_reason': 'partial_exit', 'open_date': datetime.datetime(2024, 1, 5, 23, 33, 52, 392333, tzinfo=datetime.timezone.utc), 'close_date': datetime.datetime(2024, 1, 5, 23, 34, 56, 674103, tzinfo=datetime.timezone.utc), 'stake_amount': 39.25676036292135, 'stake_currency': 'USDT', 'base_currency': 'BTC', 'fiat_currency': 'USD', 'sub_trade': True, 'cumulative_profit': -0.03887343, 'final_profit_ratio': None, 'is_final_exit': False}
2024-01-05 23:34:56,677 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=3, pair=BTC/USDT, amount=0.00089000, is_short=False, leverage=1.0, open_rate=44108.71950890, open_since=2024-01-05 23:33:52)
2024-01-05 23:34:56,681 - freqtrade.persistence.trade_model - INFO - Updating trade (id=3) ...
2024-01-05 23:34:56,681 - freqtrade.persistence.trade_model - INFO - MARKET_SELL has been fulfilled for Trade(id=3, pair=BTC/USDT, amount=0.00089000, is_short=False, leverage=1.0, open_rate=44108.71950890, open_since=2024-01-05 23:33:52).
2024-01-05 23:34:56,685 - freqtrade.wallets - INFO - Wallets synced.
2024-01-05 23:34:56,687 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': exit_fill, 'trade_id': 3, 'exchange': 'Binance', 'pair': 'BTC/USDT', 'leverage': 1.0, 'direction': 'Long', 'gain': 'loss', 'limit': 44140.95, 'order_rate': 44140.95, 'order_type': '', 'amount': 0.00044, 'open_rate': 44108.719508900394, 'close_rate': 44140.95, 'current_rate': None, 'profit_amount': -0.02464844, 'profit_ratio': -0.00126876, 'buy_tag': None, 'enter_tag': None, 'sell_reason': 'partial_exit', 'exit_reason': 'partial_exit', 'open_date': datetime.datetime(2024, 1, 5, 23, 33, 52, 392333, tzinfo=datetime.timezone.utc), 'close_date': datetime.datetime(2024, 1, 5, 23, 34, 56, 687530, tzinfo=datetime.timezone.utc), 'stake_amount': 19.848923779005176, 'stake_currency': 'USDT', 'base_currency': 'BTC', 'fiat_currency': 'USD', 'sub_trade': True, 'cumulative_profit': -0.06352187000000001, 'final_profit_ratio': None, 'is_final_exit': False}
```
The `profit_amount` is different between `exit` and `exit_fill` messages

**Comment by xmatthias** [comment]
nice investigation
care to give the new iteration a try?

**Review by stash86 — APPROVED** [review]
No more bug encountered
